<launch>

    <arg name="run_mloam" default="true" />
    <arg name="run_mloam_odom" default="true" />
    <arg name="run_mloam_mapping" default="true" />    
    <arg name="run_aloam" default="false" />
    <arg name="run_floam" default="false" />
    <arg name="run_legoloam" default="false" />

    <arg name="result_save" default="false" />
    <arg name="output_path" default="/home/jjiao/trajectory_results/real_vehicle/xxxxx" />
    <arg name="data_source" default="pcd" />
    <arg name="data_path" default="/Monster/dataset/lidar_calibration/mloam_rv_dataset/" />
    <arg name="delta_idx" default="1" />
    <arg name="start_idx" default="2100" />
    <arg name="end_idx" default="25229" />
    <arg name="with_ua" default="true" /> <!-- awareness of uncertainty propagation-->    
    <arg name="gf_method" default="wo_gf" /> <!-- gd_fix, gd_float, rnd, fps, wo_gf-->
    <arg name="gf_ratio_ini" default="1" />

    <arg name="config_file" default="$(find mloam)/config/config_realvehicle_hercules.yaml" />
    <group if="$(arg run_mloam)">
        <node pkg="mloam" type="mloam_node_rv_hercules" name="mloam_node_rv_hercules"
            args="-config_file=$(arg config_file)
                  -result_save=$(arg result_save)
                  -output_path=$(arg output_path)
                  -data_source=$(arg data_source)
                  -data_path=$(arg data_path)
                  -delta_idx=$(arg delta_idx)
                  -start_idx=$(arg start_idx)
                  -end_idx=$(arg end_idx)" required="true" output="screen">
            <remap from="/base_odom_gt" to="/current_odom"/>
        </node>
        <group if="$(arg run_mloam_mapping)">
            <node pkg="mloam" type="lidar_mapper_keyframe" name="lidar_mapper_keyframe"
                args="-config_file=$(arg config_file)
                      -result_save=$(arg result_save) 
                      -output_path=$(arg output_path)
                      -with_ua=$(arg with_ua)
                      -gf_method=$(arg gf_method)
                      -gf_ratio_ini=$(arg gf_ratio_ini)" output="screen">
                <remap from="/laser_odom" to="/laser_odom_0"/>
            </node>
        </group>
    </group>

    <!-- publish static tf -->
    <!-- x y z yaw pitch roll -->
    <!-- x y z qx qy qz qw -->
    <node name="static_tf_base_sensor_into_base_link" pkg="tf" type="static_transform_publisher"
        args="0 0 -1.5 0 0 0 1 aft_mapped base_link 10"/>                
    <node name="static_tf_world_into_world_base_sensor" pkg="tf" type="static_transform_publisher"
        args="0 0 -1.5 0 0 0 1 world world_base 10"/>
    <node name="static_tf_world_sensor_into_base_sensor" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 world laser_0 10"/>
    <node name="static_tf_laser_0_into_velo" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 laser_0 velo 10"/>        
    <node name="static_tf_laser_0_into_laser_1" pkg="tf" type="static_transform_publisher"
        args="0.5355 0.0393 -1.131 -0.0169 0.0575 0.0195 0.998 laser_0 laser_1 10"/>
    <node name="static_tf_laser_0_into_laser_2" pkg="tf" type="static_transform_publisher"
        args="-0.1118 0.1894 0.6845 0.6951 0.5116 0.6440 -0.904 laser_0 laser_2 10"/>
    <node name="static_tf_laser_0_into_laser_3" pkg="tf" type="static_transform_publisher"
        args="0.0745 0.1312 -0.7449 0.6496 0.4406 -0.628 -1.0295 laser_0 laser_3 10"/>                
 
    <!-- merge cloud -->
    <group unless="$(arg run_mloam)">
        <node pkg="mloam_test" type="test_merge_pointcloud_rv_hercules" name="test_merge_pointcloud_rv_hercules" 
            args="$(arg data_source) $(arg data_path) $(arg delta_idx) 
                  $(arg start_idx) $(arg end_idx) $(arg config_file)" required="true" output="screen"/>
        <node name="static_tf_world_sensor_into_camera" pkg="tf" type="static_transform_publisher"
            args="0 0 0 0 0 0 1 world camera_init 10"/>            
        <node name="static_tf_world_sensor_into_laser" pkg="tf" type="static_transform_publisher"
            args="0 0 0 0 0 0 1 world rslidar 10"/>
    </group>

    <!-- run aloam  -->
    <group if="$(arg run_aloam)">
        <include file="$(find aloam_velodyne)/launch/aloam_exp.launch">
            <arg name="output_path" value="$(arg output_path)"/>
            <arg name="skip_frame" value="2" />
            <arg name="s_line" value="32" />
            <arg name="mapping_line_res" value="0.2" />
            <arg name="mapping_plane_res" value="0.4" />
        </include>       
   </group>        

    <!-- run floam  -->
    <group if="$(arg run_floam)">
        <include file="$(find floam)/launch/floam_exp.launch">
            <arg name="output_path" value="$(arg output_path)"/>
            <arg name="s_line" value="32" />
            <arg name="map_res" value="0.2" />
        </include>       
   </group>        

    <!-- run legoloam  -->
    <group if="$(arg run_legoloam)">
        <include file="$(find lego_loam)/launch/legoloam_exp.launch">
            <arg name="output_path" value="$(arg output_path)"/>
        </include>
    </group>           

    <arg name="bag_file_use" default="true" />
    <group if="$(arg bag_file_use)">
        <param name="use_sim_time" value="true" />
        <node pkg="rosbag" type="play" name="rosbag" args="--clock -r 0.6 $(arg bag_file)" required="true"/>    

        <include file="$(find rviz_car_model)/launch/default.launch" >
            <arg name="fixed_frame" value="base_link" />
            <arg name="transform" value="0 0 -0.5 0 0 0 1" />
        </include>    
    </group>    
    <group unless="$(arg bag_file_use)">
        <param name="use_sim_time" value="false" />
    </group>

</launch>
